### PR TITLE
Add method to construct MeshValueCollection from arrays

### PR DESCRIPTION
**Pull request by iitrabhi**
This PR extends the functionality of MeshValueCollection class with a constructor that supports its creation from primitive arrays. The syntax is as follows:

`mvc = MeshValueCollection("size_t", mesh, dim, cells, cell_data)`

When creating a MeshValueCollection an argument specifying the type of the MeshValueCollection must be given. Allowed types are ‘int’, ‘size_t’, ‘double’ and ‘bool’. The second argument is the mesh and third is the topological dimension of the MeshValueCollection. The fourth argument is an array specifying the vertices of the mesh entity to be marked and the fifth argument is an array containing the marker of tagged entities. The length of the arrays of argument four and five must be the same. 

A unit test is defined to check the functionality of the new constructor. A demo file is also added that demonstrates the new functionality.

### CODE REVIEW

File: cpp/dolfin/mesh/MeshValueCollection.h [hunk 1]
```diff
@@ -184,6 +189,98 @@ MeshValueCollection<T>::MeshValueCollection(std::shared_ptr<const Mesh> mesh,
...
+    assert(_mesh->topology().connectivity(_dim, D));
+    const Connectivity& connectivity = *_mesh->topology().connectivity(_dim, D);
+
+    for (std::size_t j = 0; j < values_data.size();
+         ++j)
+    {
+      // Find the cell
+      v.clear();
+      // cells[j] is a vector of length _dim+1
+      for (std::size_t i = 0; i < _dim+1; ++i){
```
Comment (michalhabera) [inline cpp/dolfin/mesh/MeshValueCollection.h]: Wrong assumption on number of vertices per cell here again.

File: python/demo/mvc-from-array/demo_mvc_from_array.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+# specify Measures to integrate over specified regions and to specify
+# different boundary conditions. The above mesh value collections are defined
+# with the sole purpose of populating mesh functions.
+
+
+# ----------------Step - 4 - Construct mesh function -------------------------
+mf_triangle = cpp.mesh.MeshFunctionSizet(mesh, mvc_subdomain, 0)
+mf_line = cpp.mesh.MeshFunctionSizet(mesh, mvc_boundaries, 0)
+
+print(mf_line.values)
```
Comment (michalhabera) [inline python/demo/mvc-from-array/demo_mvc_from_array.py]: Do not leave print() statement here

File: python/test/unit/mesh/test_mesh_value_collection.py [hunk 3]
```diff
@@ -166,3 +166,40 @@ def test_mesh_function_assign_2D_vertices():
         for i in range(num_cell_vertices):
             assert 25 == g.get_value(c, i)
             assert f2.values[vertices[i]] == g.get_value(c, i)
+
+
+def test_mvc_construction_array():
```
Comment (michalhabera) [inline python/test/unit/mesh/test_mesh_value_collection.py]: Would be good to test quad/hex mesh and also MVC on vertices, can you add tests for this?

File: python/demo/mvc-from-array/demo_mvc_from_array.py (hunk 2 above)
Comment (iitrabhi) [reply python/demo/mvc-from-array/demo_mvc_from_array.py]: done

File: cpp/dolfin/mesh/MeshValueCollection.h (hunk 1 above)
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: updated

File: cpp/dolfin/mesh/MeshValueCollection.h [hunk 4]
```diff
@@ -184,6 +189,98 @@ MeshValueCollection<T>::MeshValueCollection(std::shared_ptr<const Mesh> mesh,
...
+    }
+    // The vector of vertex number is key and entity index is value
+    entity_map[v]=m.index();
+  }
+
+  const std::size_t D = _mesh->topology().dim();
+
+  // Handle cells as a special case
+  if ((int)D == _dim)
+  {
```
[the remark this answers is not in the document]
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: done

File: cpp/dolfin/mesh/MeshValueCollection.h [hunk 5]
```diff
@@ -184,6 +189,98 @@ MeshValueCollection<T>::MeshValueCollection(std::shared_ptr<const Mesh> mesh,
...
+    else
+    {
+      v.clear();
+      for (auto& vtx : mesh::EntityRange(m, 0)){
+        v.push_back(global_indices[vtx.index()]);
+      }
+      std::sort(v.begin(), v.end());
+    }
+    // The vector of vertex number is key and entity index is value
+    entity_map[v]=m.index();
```
[the remark this answers is not in the document]
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: done

File: cpp/dolfin/mesh/MeshValueCollection.h [hunk 6]
```diff
@@ -184,6 +189,98 @@ MeshValueCollection<T>::MeshValueCollection(std::shared_ptr<const Mesh> mesh,
...
+                      std::vector<std::vector<T>>& cells,
+                      std::vector<T>& values_data): _mesh(mesh), _dim(dim)
+{ 
+  // Ensure the mesh dimension is initialised. 
+  // If the mesh is created from arrays, entities 
+  // of all dimesions are not initialized.
+  _mesh->create_entities(dim);
+
+  // The number of vertices per entity is _dim+1
+  std::vector<std::int32_t> v(_dim+1);
```
[the remark this answers is not in the document]
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: done

File: cpp/dolfin/mesh/MeshValueCollection.h [hunk 7]
```diff
@@ -52,6 +52,11 @@ class MeshValueCollection
   ///         The mesh entity dimension for the mesh value collection.
   MeshValueCollection(std::shared_ptr<const Mesh> mesh, std::size_t dim);
 
+  MeshValueCollection(std::shared_ptr<const Mesh> mesh,
```
[the remark this answers is not in the document]
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: done

File: cpp/dolfin/mesh/MeshValueCollection.h [hunk 8]
```diff
@@ -52,6 +52,11 @@ class MeshValueCollection
...
+                    std::size_t dim,
+                    std::vector<std::vector<T>>& cells,//Cells
```
[the remark this answers is not in the document]
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: done

File: python/test/unit/mesh/test_mesh_value_collection.py (hunk 3 above)
Comment (iitrabhi) [reply python/test/unit/mesh/test_mesh_value_collection.py]: Working on it

Comment (iitrabhi) [reply python/test/unit/mesh/test_mesh_value_collection.py]: done

File: cpp/dolfin/mesh/MeshValueCollection.h (hunk 1 above)
Comment (iitrabhi) [reply cpp/dolfin/mesh/MeshValueCollection.h]: corrected

**Comment by michalhabera** [comment]
Reopening as a new PR.

**Pull request closed without merging**